### ISSUE TITLE
current git-master (91845da) fails to build on musl

### ISSUE THREAD

**Issue by ernsteiswuerfel** [md-raid-utilities/mdadm#76]
Get this on both of my musl boxes, musl 1.2.4 and musl 1.2.5:
```
[...]
gcc -Wall -Werror -Wstrict-prototypes -Wextra -Wno-unused-parameter -Wformat -Wformat-security -Werror=format-security -fstack-protector-strong -fPIE -Warray-bounds -Wimplicit-fallthrough=0 -Wformat-overflow -Wstringop-overflow -fno-strict-overflow -fno-delete-null-pointer-checks -fwrapv -O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE -DSendmail=\""/usr/sbin/sendmail -t"\" -DCONFFILE=\"/etc/mdadm.conf\" -DCONFFILE2=\"/etc/mdadm/mdadm.conf\" -DMAP_DIR=\"/run/mdadm\" -DMAP_FILE=\"map\" -DMDMON_DIR=\"/run/mdadm\" -DFAILED_SLOTS_DIR=\"/run/mdadm/failed-slots\" -DNO_COROSYNC -DNO_DLM -DVERSION=\"4.3-161-g91845dab\" -DVERS_DATE="\"2024-08-13\""  -DUSE_PTHREADS -DBINDIR=\"/sbin\"  -o restripe.o -c restripe.c
restripe.c: In function 'restore_stripes':
restripe.c:758:43: error: 'off64_t' undeclared (first use in this function); did you mean 'off_ '?
  758 |                                          (off64_t)read_offset) {
      |                                           ^~~~~~~
      |                                           off_t
restripe.c:758:43: note: each undeclared identifier is reported only once for each function it appears in
restripe.c:758:51: error: expected ')' before 'read_offset'
  758 |                                          (off64_t)read_offset) {
      |                                                   ^~~~~~~~~~~
      |                                                   )
restripe.c:757:36: note: to match this '('
  757 |                                 if (lseek64(source, read_offset, 0) !=
      |                                    ^
make: *** [Makefile:232: restripe.o] Error 1
```

**Comment by mtkaczyk** [comment]
Thanks!
Looks like we need to rework it. I didn't suspect that there might be a problem with availability of off64_t.
Feel free to provide a PR if you have idea how to solve it.

@ncroxon  why we need this casting? Is there issue you observed?

**Comment by ncroxon** [comment]
@mtkaczyk, I haven't changed this code nor has anyone else since 2011

758be4f1c (Lukasz Dorau       2011-09-19 13:26:05 +1000  757)                           if (lseek64(source, read_offset, 0) !=
758be4f1c (Lukasz Dorau       2011-09-19 13:26:05 +1000  758)                                    (off64_t)read_offset) {

**Comment by mtkaczyk** [comment]
Oh yes, somehow I assumed that it is regression introduced by recent patches and I remember that you added similar casting somewhere. You are right, it is old issue.

**Comment by mtkaczyk** [comment]
I looked into, and so far I see we do not set ` __USE_LARGEFILE64` and we have some weird handling of `lseek64` in mdadm.h,. Probably it is some legacy stuff not needed now but I is hard to predict (added 23 years ago).
We have also, multiple calls to `lseek` and `lseek64`. I think that the best way to fix it is:
11 Add __USE_LARGEFILE64 to force 64 bits.
2. Replace all `lseek64` by `lseek`.
3. Then problem with `off64_t` and `off_t` should be resolved, the decision will belong to complier (off_t is 32 or 64 basing on the variable mentioned above).

According to documentation setting `__USE_LARGEFILE64 ` should always force 64, seems safe?
I will propose PR after new release (I need to fix #66 first) because it is potentially dangerous.

**Comment by mtkaczyk** [comment]
@ernsteiswuerfel 
Finally, I'm mowing this forward. First, I'm trying to compile with musl and here is what I did on ubuntu 24.04
```bash
apt install musl musl-dev musl-tools

make CC=musl-gcc
```

And got following error:
```bash
mdadm.h:86:17: fatal error: linux/kdev_t.h: No such file or directory
   86 | #include        <linux/kdev_t.h>               
```
Could you please support me? I would love to merge PR with musl compilation, even if it is failing now.

**Comment by mtkaczyk** [comment]
The one way to make at least musl compilation always reliable is to have it tested here, in mdadm repo. Otherwise we may break it in any moment because we didn't notice. I think we should support musl but that require both of us work together. Could you please help?

**Comment by mtkaczyk** [comment]
I spent some time on it and determined that I need to provide `/usr/include`. So the command I used is:
```bash
make CC=musl-gcc CXFLAGS="-idirafter /usr/include"
```
but it is not working because musl is not providing `asm/` headers. We are importing `asm/types.h` in mdadm.
I don't see it possible to compile.
```
musl-dev/noble,now 1.2.4-2 amd64 [installed]
musl-tools/noble,now 1.2.4-2 amd64 [installed]
musl/noble,now 1.2.4-2 amd64 [installed]
```

Looks like musl version is recent. Any suggetions? Thanks in advance!

**Comment by ernsteiswuerfel** [comment]
> Could you please support me? I would love to merge PR with musl compilation, even if it is failing now.

Sure thing I will try to help out! 🤓 

On my Gentoo musl/llvm box it builds for a while and then I still get:
```
 # make CC=gcc
gcc -Wall -Werror -Wstrict-prototypes -Wextra -Wno-unused-parameter -Wformat -Wformat-security -Werror=format-security -fstack-protector-strong -fPIE -Warray-bounds -Wimplicit-fallthrough=0 -Wformat-overflow -Wstringop-overflow -fno-strict-overflow -fno-delete-null-pointer-checks -fwrapv -O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE -DSendmail=\""/usr/sbin/sendmail -t"\" -DCONFFILE=\"/etc/mdadm.conf\" -DCONFFILE2=\"/etc/mdadm/mdadm.conf\" -DMAP_DIR=\"/run/mdadm\" -DMAP_FILE=\"map\" -DMDMON_DIR=\"/run/mdadm\" -DFAILED_SLOTS_DIR=\"/run/mdadm/failed-slots\" -DNO_COROSYNC -DNO_DLM -DVERSION=\"4.4-24-g7d29b382\" -DVERS_DATE="\"2025-03-03\""  -DUSE_PTHREADS -DBINDIR=\"/sbin\"  -o restripe.o -c restripe.c
restripe.c: In function 'restore_stripes':
restripe.c:760:43: error: 'off64_t' undeclared (first use in this function); did you mean 'off_t'?
  760 |                                          (off64_t)read_offset) {
      |                                           ^~~~~~~
      |                                           off_t
restripe.c:760:43: note: each undeclared identifier is reported only once for each function it appears in
restripe.c:760:51: error: expected ')' before 'read_offset'
  760 |                                          (off64_t)read_offset) {
      |                                                   ^~~~~~~~~~~
      |                                                   )
restripe.c:759:36: note: to match this '('
  759 |                                 if (lseek64(source, read_offset, 0) !=
      |                                    ^
make: *** [Makefile:232: restripe.o] Error 1
```
This is from git-7d29b3823c18a24d6efbb502f08638788f97e04b.

On Gentoo the mdadm-4.4.ebuild uses a workaround:
```
[...]
src_compile() {
        use static && append-ldflags -static

        # CPPFLAGS won't work for this
        use udev || append-cflags -DNO_LIBUDEV

        # bug 907082
	use elibc_musl && append-cppflags -D_LARGEFILE64_SOURCE

        mdadm_emake all
}
[...]
```

**Comment by ernsteiswuerfel** [comment]
> I spent some time on it and determined that I need to provide `/usr/include`. So the command I used is:
> 
> make CC=musl-gcc CXFLAGS="-idirafter /usr/include"
> 
> but it is not working because musl is not providing `asm/` headers. We are importing `asm/types.h` in mdadm. I don't see it possible to compile.

Can't tell you much about building with musl on Distros with another libc installed unfortunately...

Gentoo uses following logic in the ebuild:
```
[...]
DEPEND="
        udev? ( virtual/libudev:= )
        corosync? ( sys-cluster/corosync )
"
RDEPEND="
	${DEPEND}
        >=sys-apps/util-linux-2.16
"

# The tests edit values in /proc and run tests on software raid devices.
# Thus, they shouldn't be run on systems with active software RAID devices.
RESTRICT="test"

PATCHES=(
	"${WORKDIR}/debian/patches/debian"
)

mdadm_emake() {
        # We should probably make libdlm into USE flags (bug #573782)
        local args=(
                PKG_CONFIG="$(tc-getPKG_CONFIG)"
                CC="$(tc-getCC)"
                CWFLAGS="-Wall -fPIE"
                CXFLAGS="${CFLAGS}"
                LDFLAGS="${LDFLAGS}"
                UDEVDIR="$(get_udevdir)"
                SYSTEMD_DIR="$(systemd_get_systemunitdir)"
                COROSYNC="$(usev !corosync '-DNO_COROSYNC')"
                DLM="-DNO_DLM"

                # bug #732276
                STRIP=

                "$@"
        )

	emake "${args[@]}"
}
[...]
```

> ```
> musl-dev/noble,now 1.2.4-2 amd64 [installed]
> musl-tools/noble,now 1.2.4-2 amd64 [installed]
> musl/noble,now 1.2.4-2 amd64 [installed]
> ```
> 
> Looks like musl version is recent. Any suggetions? Thanks in advance!

musl 12.4. was recent enough to provoke the original issue so it should do. 😉 Gentoo uses 1.2.5 meanwhile.

**Comment by mtkaczyk** [comment]
@ernsteiswuerfel 
Just curious how you providing `asm/types.h` in your system. On my docker ubuntu image I have:

```
echo | musl-gcc -idirafter /usr/include -E -Wp,-v -
ignoring duplicate directory "/usr/include/x86_64-linux-musl"
ignoring duplicate directory "/usr/lib/gcc/x86_64-linux-gnu/13/include"
#include "..." search starts here:
#include <...> search starts here:
 /usr/include/x86_64-linux-musl
 /usr/lib/gcc/x86_64-linux-gnu/13/include
 /usr/include
End of search list.
```

for glibc I have it as follows:
```
echo | gcc -E -Wp,-v -
ignoring nonexistent directory "/usr/local/include/x86_64-linux-gnu"
ignoring nonexistent directory "/usr/lib/gcc/x86_64-linux-gnu/13/include-fixed/x86_64-linux-gnu"
ignoring nonexistent directory "/usr/lib/gcc/x86_64-linux-gnu/13/include-fixed"
ignoring nonexistent directory "/usr/lib/gcc/x86_64-linux-gnu/13/../../../../x86_64-linux-gnu/include"
#include "..." search starts here:
#include <...> search starts here:
 /usr/lib/gcc/x86_64-linux-gnu/13/include
 /usr/local/include
 /usr/include/x86_64-linux-gnu
 /usr/include
End of search list.
```
`asm/types.h` are defined in libc `usr/include/x86_64-linux-gnu/asm/types.h` but the code there is simple we just do:
```
#include <asm-generic/types.h>
```
In `/usr/include/x86_64-linux-musl/asm`  there is no definition of `types.h`. Looks like you have it solved in you build system, could you please dig a little and find how? :)

**Comment by ernsteiswuerfel** [comment]
**/usr/include/asm/types.h:**
```
 # equery b /usr/include/asm/types.h
 * Searching for /usr/include/asm/types.h ... 
sys-kernel/linux-headers-6.12 (/usr/include/asm/types.h)
 # cat /usr/include/asm/types.h 
#include <asm-generic/types.h>
```

**/usr/include/asm-generic/types.h**
```
 # equery b /usr/include/asm-generic/types.h
 * Searching for /usr/include/asm-generic/types.h ... 
sys-kernel/linux-headers-6.12 (/usr/include/asm-generic/types.h)
 # cat /usr/include/asm-generic/types.h 
/* SPDX-License-Identifier: GPL-2.0 WITH Linux-syscall-note */
#ifndef _ASM_GENERIC_TYPES_H
#define _ASM_GENERIC_TYPES_H
/*
 * int-ll64 is used everywhere now.
 */
#include <asm-generic/int-ll64.h>

#endif /* _ASM_GENERIC_TYPES_H */
```

**/usr/include/sys/types.h:**
```
 # equery b /usr/include/sys/types.h
 * Searching for /usr/include/sys/types.h ... 
sys-libs/musl-1.2.5-r3 (/usr/include/sys/types.h)
 # cat /usr/include/sys/types.h
#ifndef	_SYS_TYPES_H
#define	_SYS_TYPES_H
#ifdef __cplusplus
extern "C" {
#endif

#include <features.h>

#define __NEED_ino_t
#define __NEED_dev_t
#define __NEED_uid_t
#define __NEED_gid_t
#define __NEED_mode_t
#define __NEED_nlink_t
#define __NEED_off_t
#define __NEED_pid_t
#define __NEED_size_t
#define __NEED_ssize_t
#define __NEED_time_t
#define __NEED_timer_t
#define __NEED_clockid_t

#define __NEED_blkcnt_t
#define __NEED_fsblkcnt_t
#define __NEED_fsfilcnt_t

#define __NEED_id_t
#define __NEED_key_t
#define __NEED_clock_t
#define __NEED_suseconds_t
#define __NEED_blksize_t

#define __NEED_pthread_t
#define __NEED_pthread_attr_t
#define __NEED_pthread_mutexattr_t
#define __NEED_pthread_condattr_t
#define __NEED_pthread_rwlockattr_t
#define __NEED_pthread_barrierattr_t
#define __NEED_pthread_mutex_t
#define __NEED_pthread_cond_t
#define __NEED_pthread_rwlock_t
#define __NEED_pthread_barrier_t
#define __NEED_pthread_spinlock_t
#define __NEED_pthread_key_t
#define __NEED_pthread_once_t
#define __NEED_useconds_t

#if defined(_GNU_SOURCE) || defined(_BSD_SOURCE)
#define __NEED_int8_t
#define __NEED_int16_t
#define __NEED_int32_t
#define __NEED_int64_t
#define __NEED_u_int64_t
#define __NEED_register_t
#endif

#include <bits/alltypes.h>

#if defined(_GNU_SOURCE) || defined(_BSD_SOURCE)
typedef unsigned char u_int8_t;
typedef unsigned short u_int16_t;
typedef unsigned u_int32_t;
typedef char *caddr_t;
typedef unsigned char u_char;
typedef unsigned short u_short, ushort;
typedef unsigned u_int, uint;
typedef unsigned long u_long, ulong;
typedef long long quad_t;
typedef unsigned long long u_quad_t;
#include <endian.h>
#include <sys/select.h>
#endif

#if defined(_LARGEFILE64_SOURCE)
#define blkcnt64_t blkcnt_t
#define fsblkcnt64_t fsblkcnt_t
#define fsfilcnt64_t fsfilcnt_t
#define ino64_t ino_t
#define off64_t off_t
#endif

#ifdef __cplusplus
}
#endif
#endif
```

**Comment by ernsteiswuerfel** [comment]
Gentoos' `sys-kernel/linux-headers` are simply the headers from https://www.kernel.org/ in an own package:
https://packages.gentoo.org/packages/sys-kernel/linux-headers
